Add Assure() as a replacement for problematic Must()

rousskov opened the pull request:
The now-deprecated Must() has several interdependent flaws:

1. Must() logs at debug level 3, hiding some important bugs from humans.

2. Must() has been (ab)used for both checking code logic and validating
   input, making purpose-specific implementation changes impractical.

3. Must() does not honor the standard NDEBUG macro, complicating runtime
   cost assessment and surprising some developers that are used to that
   standard assert() semantics.

The new Assure() is a throwing version of a POSIX assert(3):

* Meant for detecting Squid code logic bugs (not input validation).
* Informs admins about bugs by logging their info to cache.log.
* Completely disabled in NDEBUG builds.
* Kills the current component (e.g., a Squid-origin HTTP transaction).

The killed component boundary is essentially defined by the location of
the handling try/catch statement. Some Assure() failures will kill a
Squid process, but the throwing code should not worry (or even know)
about the catcher location and handling logic.

This change also optimizes the compiled Must()/Assure() caller code
size, which may help a bit with runtime performance: With the new
Assure()/Must() implementation approach, the total stripped Squid
executable size in one reasonable configuration goes down by 5%. For
comparison, removing all Must()s completely gives 6% size reduction.

Automatically replacing Must() calls with Assure() is not practical due
to the second flaw itemized above.

rousskov commented:
The new Assert() macro is meant for situations like the one discussed in https://github.com/squid-cache/squid/pull/795#discussion_r606503795 . Many assert() calls and most Must() calls should be eventually replaced with Assert(). I also want to use it in several upcoming PRs. Please see https://github.com/measurement-factory/squid-notes/blob/start/errors.md#invariant-checks-missing-api for the extended rationale.

yadij commented on src/base/Assert.h:
```diff
@@ -0,0 +1,52 @@
+/// the corresponding exception. Reduces compiled code size of Assert() and
+/// Must() callers. Do not call directly; use Assert() instead.
+/// \param description explains the condition (i.e. what MUST happen)
+[[ noreturn ]] void ReportAndThrow_(int debugLevel, const char *description, const SourceLocation &);
```
"report" how to where, etc ?
```suggestion
[[ noreturn ]] void LogAndThrow_(int debugLevel, const char *description, const SourceLocation &);
```

yadij reviewed (changes requested):
I see quite a few issues with this PR as-written. To begin with lets settle the design issues.

An API re-inventing a slightly non-standard copy of stdlib assert() is a bad idea. This is exactly the same approach taken with Must(), which led to the issues complained about.
 ... this time around please use self-documenting names for the API functions and provide wrappers for *both* code validation and input validation.

yadij commented:
> > An API re-inventing a slightly non-standard copy of stdlib assert() is a bad idea.
> 
> Agreed. This PR does not reinvent a slightly non-standard copy of assert(). As I painstakingly detailed [elsewhere](https://github.com/measurement-factory/squid-notes/blob/start/errors.md#invariant-checks-missing-api), the scope of PR Assert() effect is different from that of the standard assert(): The standard assert() kills the current kid process. PR Assert() terminates the current processing component (which will usually be a lot smaller than the entire kid process) and triggers orderly stack unwinding. That difference is far from "slight"!

The difference is only in its side-effect. That is slight enough to be confusing to anyone new reading our code, and large enough to be bad when they get it wrong. I am strongly against using the word "assert" to name (or document) for things that are not stdlib assert().

rousskov commented:
> I am strongly against using the word "assert" to name (or document) for things that are not stdlib assert().

Please propose a better name.

rousskov commented:
You have [suggested](https://github.com/squid-cache/squid/pull/864#discussion_r673626835) to rename a private helper function to LogAndThrowIf(). If I correct that proposal and apply the corrected version to the public API in question, I get something like LogAndThrowIfNot() or LogAndThrowUnless(). I sincerely hope that is _not_ the name you think is better than Assert(), so I am asking for explicit suggestions that would allow us to make progress with this important PR.

rousskov commented:
> > Amos: I am strongly against using the word "assert" to name (or document) for things that are not stdlib assert().

> Alex: Please propose a better name.

We can use Require(), Insist(), Vow() or even Assume() instead of Assert(). @yadij, would one of those names address your concern? There is also Posit() and Avow() but they will be unfamiliar to a larger share of non-native speakers.

kinkie commented:
I support this plan. It is clearer than what we have now

rousskov commented:
Unfortunately, the proposed [plan](https://github.com/squid-cache/squid/pull/864#issuecomment-1014816977) does not address the problem this PR is addressing. If you have not seen the research that went into this PR, please read [it](https://github.com/measurement-factory/squid-notes/blob/start/errors.md#invariant-checks-missing-api).

In summary, to start fixing the stated problems with Must(), we need a new function that is used for bug detection (like assert()) but only kills the current component (like Must()). The functions proposed in the plan are not it:

* Requires() is dedicated to input validation[^1]. The PR-proposed Assert() is _not_ for input validation. It is for bug detection. 
* Fatal() kills the current process. The PR-proposed Assert() is for killing the current component (e.g., a master transaction). It is also impractical to implement Fatal() via exceptions in Squid, so Fatal() is essentially the existing abort()-calling assert().

If the plan is implemented, we will still have to pick between Must() and assert() when detecting component implementation bugs, and that choice is always wrong. This is the problem this PR is addressing.

[^1]: When doing input validation, in most cases, we do not want a general-purpose Must()-like global function. We want context-specific throwing code that can appropriately _detail_ the input problem rather than just sharing the parser source code location. That context-specific throwing code may look like `require(condition, input_description)`, but that is a different context-specific beast, with different needs (e.g., support for nested contexts). It is a whole separate known problem worth discussing/solving, but independent from this PR.

rousskov commented:
@kinkie, the ball is on your side to name the new macro. Please treat this PR as a priority because a large variety of pending work would benefit from getting this improvement merged. You only need to suggest a name or two you like the best -- once we reach consensus on a specific spelling, I will do the legwork of applying the changes to the PR code and metadata.

FWIW, here is the list of macro names [updated](https://github.com/squid-cache/squid/pull/864#issuecomment-888757965) with more options and to remove my first alternative choice -- the currently reserved Require() -- as we agreed during the conference call. You do not have to pick from _this_ list, of course!

* Insist()
* Vow()
* Uphold()
* Confirm()
* Verify()
* Validate(): Too close to the "input validation" phrase that is not applicable to this macro?
* ThrowUnless(): I would only use this as a last resort because it is _too specific_  -- the macro does _not_ throw in `NDEBUG` builds (just like assert() does not abort() in those builds). Something called "throw unless `c`" should not have throwing preconditions other than `!c` itself. The "unless c" structure is also more difficult for most humans to interpret correctly.
* Posit(): May look unfamiliar to many developers with weak English skills?
* Avow(): May look unfamiliar to many developers with weak English skills?
* Assume(): MSVS has [__assume()](https://docs.microsoft.com/en-us/cpp/intrinsics/assume?view=msvc-170) with somewhat different purpose/semantics (that better match this name than our macro would); the differences may confuse folks familiar with MS compilers. I would avoid this one.

IIRC, you were concerned about name clashes with other libraries. That concern is valid. However, I would not worry too much about it because most libraries are not going to declare macros like that -- most use lower-case names, all-caps names, or library-name-prefixed names. Must() has not clashed. And if there is a clash, it is not the end of the world -- we can easily rename our macro with a well-chosen name to something else if needed. Thus, I would not add words to the macro name _just_ to avoid clashes: Some Verb() spellings are better than their more verbose VerbThat() variations.

kinkie commented:
@rousskov , I hope I'm not sending this conversation back to the drawing board too much so I won't insist on rethinking about designed choices, but I wonder: code correctness is a big deal; should we even honor NODEBUG?

This would mean that we could just keep ReportAndThrow (LogAndThrow?) as the main API point, possibly by turning it from a macro into a static inline. Given that it should never really assert, branch predictors would have an easy time in taking the common path (we might even hint it in the code), so the runtime cost should be really minimal

What do you think?

rousskov commented:
> I hope I'm not sending this conversation back to the drawing board too much, but I wonder should we even honor NDEBUG?

I think this is a valid question[^1]. I believe the answer is "yes": If we can easily support a very cheap and potentially useful standard optional feature, we should support it. Here are some of the factors that contributed to that "yes" answer:

1. `NDEBUG` is a [standard](https://en.cppreference.com/w/cpp/error/assert) C++ feature applicable to this context.
2. The feature is disabled by default. If admin wants to enable it[^2], we should trust that admin.
3. The feature state is tested at compile time so supporting this on/off switch does not hurt runtime performance.
4. We already explicitly support (the misspelled[^1] version of) `NDEBUG` in an equivalent context of assert().
5. Honoring `NDEBUG` answers the question whether assertions (of all kinds) may contain code that does more than just checking the current state. The answer clearly becomes "no". Squid currently violates that principle in a few buggy assert() statements, but those temporary bugs are not important in this context.
6. The feature is useful for testing performance overhead of assertions (of all kinds). If Squid performance does not significantly improve after assertions are disabled, then admins would be less tempted to disable assertions[^2], and developers would be less tempted to avoid adding useful assertions. Many years from now, `-DNDEBUG` may even become one of the regular performance CI tests!


[^1]: I have fixed the macro spelling and removed the "code correctness" "loading" of the question to be able to say that :-). Yes, Squid code uses `NODEBUG` instead of `NDEBUG`, but I bet that is a bug/typo added in commit edd2eb6 dedicated to helping Purify. I have seen other/independent references on the web that mention Purify-friendly software builds defining `NDEBUG`.

[^2]:  In the future, _after_ PR-unrelated assert(3) bugs are fixed. There is still buggy assert(doMoreThanJustChecking()) code like `assert(e.mem_obj->data_hdr.write(buf))` that makes `NDEBUG` a bad idea _today_, but we will fix that eventually.

 
> This would mean that we could just keep ReportAndThrow (LogAndThrow?) as the main API point,

FWIW, those names do not work well for something that does not usually report/log and does not usually throw. You would have to call it ReportAndThrowUnless(condition).

I understand that I [discouraged](https://github.com/squid-cache/squid/pull/864#issuecomment-1024524748) you from suggesting ThrowUnless-kind of names because of `NDEBUG`. However, if this is indeed the best name others would not veto, I would not veto it just because it is misleading in `NDEBUG` builds. It would be a very sad day when vetoes block all reasonable naming options and leave us with ReportAndThrowUnless(), but merging this PR with ReportAndThrowUnless() is still better than doing nothing IMO, regardless of `NDEBUG` support.

> possibly by turning it from a macro into a static inline.

We need a macro to collect `file:line` information the right way: Explicitly adding Here() to every `ReportAndThrowUnless(c, Here())` call is wrong for both performance and aesthetic reasons. Unlike explicit exception throwing code that is _naturally_ verbose and slow (and, hence, should use explicit Here() instead of macro tricks), asserts (of all kinds) are meant to be as fast and as succinct as possible. Otherwise, folks will simply not add them when they should.

> Given that it should never really assert, branch predictors would have an easy time in taking the common path (we might even hint it in the code), so the runtime cost should be really minimal

I agree that an optimizing compiler should be able to use a simple inlined function just like a macro as far as path prediction is concerned. I am not sure Coverity is that powerful/flexible, but all that does not matter because asserts (of all kinds) cannot use a function for other reasons.

yadij commented:
> > I hope I'm not sending this conversation back to the drawing board too much, but I wonder should we even honor NDEBUG?
> 
> I think this is a valid question[1](#user-content-fn-1-b4b328a8300a047fe54d53ce8db1a066). I believe the answer is "yes": If we can easily support a very cheap and potentially useful standard optional feature, we should support it. Here are some of the factors that contributed to that "yes" answer:
> 

On the other hand I believe the answer is "no" - for the same reasons you stated AND the "assert is POSIX assert(2)" reason.


>     1. `NDEBUG` is a [standard](https://en.cppreference.com/w/cpp/error/assert) C++ feature applicable to this context.
> 

The "applicable to this context" is debated per the definition of "this context" being Squid code wrapping ```throw``` statements.

It is relevant to code which calls assert(2). In which cases that code is being fixed to actually call assert(), not throw.

So, neither case actually needs us to write NDEBUG in the caller code.


>     2. The feature is disabled by default. If admin wants to enable it[2](#user-content-fn-2-b4b328a8300a047fe54d53ce8db1a066), we should trust that admin.
> 

AFAIK the feature is routinely enabled in "production" builds. see also (3)


>     3. The feature state is tested at compile time so supporting this on/off switch does not hurt runtime performance.
> 

Per the core dev discussion we are using exceptions for *validation* now, not state/code correctness.  Supporting NDEBUG to eliminate **input validation** moves it firmly into the security breach area and is not something I can agree with.

NP: "trust the admin" does not work in security because that requires the admin have unreasonable levels of knowledge about other actors capabilities or hidden information.


>     4. We already explicitly support (the misspelled[1](#user-content-fn-1-b4b328a8300a047fe54d53ce8db1a066) version of) `NDEBUG` in an equivalent context of assert().
> 

Exactly why this new/changed code does not need to have explicit mention of NDEBUG.

IIRC we agreed that assert(2) should be (always?) used for things which need to check state conformance. When the relevant code lines are fixed to call assert() instead of Must() the NDEBUG support pops into existence implicitly. No specific NDEBUG code mentions needed.

I think ^^that is @kinkie's point.


>     5. Honoring `NDEBUG` answers the question whether assertions (of all kinds) may contain code that does more than just checking the current state. The answer clearly becomes "no". Squid currently violates that principle in a few buggy assert() statements, but those temporary bugs are not important in this context.
> 

IMO if you are seriously considering NDEBUG in code it should probably be using assert() instead of Must(), or the reverse. Either way that code is probably wrong.


>     6. The feature is useful for testing performance overhead of assertions (of all kinds). If Squid performance does not significantly improve after assertions are disabled, then admins would be less tempted to disable assertions[2](#user-content-fn-2-b4b328a8300a047fe54d53ce8db1a066), and developers would be less tempted to avoid adding useful assertions. Many years from now, `-DNDEBUG` may even become one of the regular performance CI tests!
> 

This implies you are conflating assert() and Must(). Which highlights why this clear definition split of assert(2) calls vs Must() calls @kinkie is working on is necessary

rousskov commented:
TLDR: @yadij, I think you are misinterpreting what @kinkie and I were discussing. You may also be misunderstanding what this PR is about, and what we have (not) agreed on during the core meeting. I will try to correct some of those misunderstandings, but the next round should probably be via chat/live discussion. For that chat/live discussion setup, please see my email titled "Weekly Squid meeting" sent around January 30, 2022. We can also do that on Zulip/Zoom right now if you are around.

>> 1. `NDEBUG` is a [standard](https://en.cppreference.com/w/cpp/error/assert) C++ feature applicable to this context.
 
> The "applicable to this context" is debated per the definition of "this context" being Squid code wrapping `throw` statements.

We can obviously debate whether ThrowUnless() should support `NDEBUG`, but that debate will not change the facts that both assert() and ThrowUnless() are assertions (i.e. they are designed to assert code correctness), and that `NDEBUG` is a standard feature that controls standard assertions. The combination of those facts makes `NDEBUG` a standard feature applicable to the context of this discussion. Nothing more. Nothing less. I see no value in debating _applicability_. Applicability does not dictate any decisions. It can be used in support of a "yes" decision, but it certainly does not require that decision. We can still decide "no" for some other valid reason.

 
>> 2. The feature is disabled by default. If admin wants to enable it, we should trust that admin.
 
> AFAIK the feature is routinely enabled in "production" builds.

I am not sure what exactly you mean by "enabled" in this context, but I see no signs that Squids are build with disabled assert()s in "production" builds. Moreover, I know that most SMP Squids _cannot function_ with disabled assert()s today due to Squid bugs.


>> 3. The feature state is tested at compile time so supporting this on/off switch does not hurt runtime performance.
 
> Per the core dev discussion we are using exceptions for _validation_ now, not state/code correctness.

This PR and the future ThrowUnless() is about code state/logic validating assertions, not input validation. It has never been about input validation. It will never be about input validation. All for the reasons I have carefully detailed already. I am not aware of any discussion that would indicate otherwise, but let's re-discuss as needed.


> NP: "trust the admin" does not work in security because that requires the admin have unreasonable levels of knowledge about other actors capabilities or hidden information.

In the context of this PR, the admin will have reasonable level of understanding of the effects: If Squid is buggy, and there is an assertion (of any kind) that catches the bug, disabling assertions leads to undefined behavior. Anything can happen. From crashes, to security violations, to content corruption, to nothing at all.


> IIRC we agreed that assert(2) should be (always?) used for things which need to check state conformance. 

Both assert() and ThrowUnless(). As I have [documented](https://github.com/measurement-factory/squid-notes/blob/start/errors.md#invariant-checks-exceptions-vs-assertions), it is wrong to rely on assert() alone for state/logic correctness checks.

> > 6. The feature is useful for testing performance overhead of assertions (of all kinds). If Squid performance does not significantly improve after assertions are disabled, then admins would be less tempted to disable assertions[2](#user-content-fn-2-b4b328a8300a047fe54d53ce8db1a066), and developers would be less tempted to avoid adding useful assertions. Many years from now, `-DNDEBUG` may even become one of the regular performance CI tests!
 
> This implies you are conflating assert() and Must(). Which highlights why this clear definition split of assert(2) calls vs Must() calls @kinkie is working on is necessary

I do not conflate those different features. I have painstakingly [detailed](https://github.com/measurement-factory/squid-notes/blob/start/errors.md#invariant-checks-missing-api) assert(), Must(), and ThrowUnless() (a.k.a. Assert()) differences, similarities, and (ab)use cases and re-discussed those points on multiple occasions. I am as far from conflation here as it gets!.. 

For logic correctness checks (_not_ input validation), we need assert() _and_ ThrowUnless(). We do not want Must(), but we cannot get rid of existing Must()s quickly enough so we cannot reuse that name.

rousskov commented:
@kinkie, FYI: Amos and I discussed `NDEBUG`. See Zulip thread if you are interested in details. AFAICT, Amos will not veto the proposed `NDEBUG` support in the new macro (in principle -- the actual code still needs to be polished and reviewed, of course!). We did not have enough time to discuss other aspects of this PR, but I still think it is important to get this feature in Squid, and I hope you will find the time to propose the best name for the new macro. Thank you.

rousskov commented:
@kinkie, I have added a few more words to the list of suggestions at https://github.com/squid-cache/squid/pull/864#issuecomment-1024524748. Is there anything I can do to help you move this PR forward?

kinkie commented:
Thanks Alex for this very comprehensive list and sorry if I wasn't super responsive - travel.

My thoughts:
- I would avoid the most directive names (Require, Must, MUST, ReallyMust) as I associate them more with unrecoverable errors.
- Confirm, Should send too weak a message
- Vow, Posit, Avow - I agree that not everyone would understand them
- Assume -  besides the MSVS one, it feels to me more like something that asserts a state for testing purposes, instead of validating a runtime state
- Verify, Validate - too close to input validation, let's reserve them

I'll throw one more in the mix:
- Ensure (or its sibling EnsureThat)

I think ti conveys the right mix of assertiveness with an escape hatch (or it might be just Not Invented Here Syndrome)

What do you think? If it doesn't meet your favour, I'd choose Uphold, and finally ThrowUnless

rousskov commented:
> I would avoid the most directive names (Require, Must, MUST, ReallyMust) as I associate them more with unrecoverable errors.

The error is indeed not recoverable as far as the throwing component is concerned. That component is doomed. There may be some workarounds at a higher level, if something there will instantiate a new component or find a replacement to try again, but "non-recoverable error" is exactly the right message here IMO.

There is another way to illustrate the same idea: Do you think assert() is unrecoverable? Well, it is not: Squid may start a replacement kid/helper process and keep going! And even if we instrument assert() to really, really, kill all SMP processes, the systemd may restart the failed instance. Etc., etc.

In a sense, there are no unrecoverable errors from a "global enough" point of view -- life always goes on... All unrecoverable errors are only from their component point of view. For the assert(), that component currently is "process". For ThrowUnless(), that component could be a transaction or a similar logical thread that is usually (but not always) smaller than the process.

> * Confirm, Should send too weak a message

I think Confirm does not have that specific flaw. I agree regarding Should.
 
>  * Ensure (or its sibling EnsureThat)

FWIW, I did not include Ensure() on my list because the code cannot make sure that the condition is true. The code throws if the condition is false, but that does not make the condition true! In other words, "ensure", to me, sounds more like "do something to make sure the condition becomes true if it is not true already" and that "do something" part is not the signal we want to send. The function does _not_ have that side effect.

There is a similar word: Assure. It has less of a "make sure" connotation AFAICT.

I am not hearing it myself, but another person I asked suggested that Assure and Uphold (below) have the same "implies some action will be taken to make the condition true" problem.


> If it doesn't meet your favour, I'd choose Uphold, and finally ThrowUnless

I would order the (possibly) remaining ones as follows (from best to worse, roughly!) but the choice is _yours_. I can live with any of these, even ThrowUnless() if really necessary.

* Require(c > 0)
* Confirm(c > 0)
* Assure(c > 0)
* Uphold(c > 0)
* Ensure(c > 0)
* ThrowUnless(c > 0)

kinkie commented:
Let's go with Assure()

-- 
    Francesco

rousskov commented:
> * Require(c > 0)
> * Confirm(c > 0)
> * Assure(c > 0)
> * Uphold(c > 0)
> * Ensure(c > 0)
> * ThrowUnless(c > 0)

> Franceso: Let's go with Assure()

@yadij, is Assure() OK with you?

yadij commented:
I have no issues with Assure()

rousskov commented:
Thank you @kinkie  and @yadij for moving this important PR forward! I replaced Assert() with Assure(), polished a few comments, and updated the code to use commit d816f28 labeling for level-0/1 messages.

@yadij, I believe all your (known to me) concerns have been addressed. Please re-review while keeping in mind that

* Assure() adds a code logic assertion; the macro is _not_ meant for input validation. Assure() is like assert(3) that throws instead of calling abort(). Reviewers will reject code using Assure() for input validation, just like we already do with assert().
* Adding a general-purpose input validation macro is probably a bad idea and can be definitely considered to be outside this PR scope. This PR does not split Must() into two macros. It splits Must() into Assure() and "throw explicitly". We can debate the merits of a general purpose input validation macro separately, without blocking progress.
* I am happy to continue discussing `NDEBUG` usage in hope to crystallize your concerns so that they can be clearly addressed, one way or another. I think we agreed that there is no practical/good reason to block this PR while waiting for those discussions to resume. In the unlikely event we decide that Assure() should not know about `NDEBUG`, we can easily remove that knowledge from Assure() implementation without any affect on the existing (at the time) Assure()-calling code.
* I am aware that you did not like `BUG` sub-labels preserved by commit d816f28. I am happy to discuss your concerns, but that should be done outside this PR because the problem (if any) is not specific to this PR. This PR uses `BUG` for consistency with the existing code -- this is how modern Squid code labels non-assert() assertion (i.e. code logic) failures. If we decide that `BUG` labels should be dropped/changed, we will do that in _all_ affected places, including Assure(). That should not be done in this PR.

yadij reviewed (dismissed):
I think we should go for a fully correct API definition for Assure() here without the hacks to allow compatibility with deprecated Must().

rousskov commented on src/base/TextException.h:
```diff
@@ -57,20 +58,18 @@ std::ostream &operator <<(std::ostream &, const TextException &);
+/// Like Assure() but only logs the exception if level-3 debugging is enabled
+/// and runs even when NDEBUG macro is defined. Deprecated: Use Assure() for
+/// code logic checks and throw explicitly when input validation fails.
+#define Must(condition) \
+    Assure_(3, (condition), #condition, Here())
```
> The removal is coming as a followup PR soon

I doubt we would be able to remove ~160 Must()s soon enough to justify code duplication here. Must() removal cannot be automated (because we cannot automatically distinguish code logic checks from input validation) and many existing input-validating Must()s deserve non-trivial code replacements.

On the other hand, if Must()s are indeed removed soon, I would be very pleased and will, of course, promptly volunteer the corresponding Assure() adjustments.

Either way, the outcome will be a good one!

yadij replied:
TL;DR - I do insist on this change. We will have to agree to disagree on our predictions of the future.

Long version;
Squid already has those Must() in existence. So this request is a matter of limiting the improvement/change to the set which can be converted to Assure(). To avoid the scope-creep of fixing code unrelated to Assure() itself.

(I see you here again falling for the trap of bundling multiple code improvements: Must() removal + protocol handling improvements. The total is great long-term, very tempting. But keeping the parts independent makes it easier to get the important parts of both done quicker.)

rousskov replied:
> I do insist on this change. 

You made the right call, thank you. Please unblock this PR.

> Squid already has those Must() in existence. So this request is a matter of limiting the improvement/change to the set which can be converted to Assure(). To avoid the scope-creep of fixing code unrelated to Assure() itself.

I still see no good reason to duplicate this functionality, and I see no scope creep in this PR. I see in-scope refactoring of Must() _implementation_ to avoid significant (and costly) code/functionality duplication. This refactoring is the best way forward here IMO. Duplication is one of the primary sources of bugs and development problems in Squid. We should avoid making things worse. This PR avoids increasing duplication.

And just to clarify if somebody is reading this exchange too quickly: Leaving Must() "as is" in this PR will _not_ "limit improvements to the set of Must()s that will be converted to Assure()". Leaving Must() as is will leave _all_ Must()s "as is", including those Must()s that will be eventually converted to Assure(). Bugs notwithstanding, this PR makes the existing Squid code better, even though the new Assure() is still unused. That is a good thing! The alternative would make Squid code (slightly) worse (until we start using Assure()).

> I see you here again falling for the trap of bundling multiple code improvements

I do not see tightly-related improvements as a trap. In this specific case, I see them as the best way to make progress.

rousskov replied:
> I do insist on this change.

> You made the right call, thank you. Please unblock this PR.

Ah, I misread. Sorry. I see no valid reason for you to insist that Must() and Assure() implementations must duplicate each other. Can you give one?

And given for how long this PR was mishandled, perhaps you can find a way to let it merge despite your reservations?

yadij replied:
> Ah, I misread. Sorry. I see no valid reason for you to insist that Must() and Assure() implementations must duplicate each other. Can you give one?
> 

I am **not** insisting that. I am requesting that the existing Must() definition be left **untouched** by this PR.

AFAICS, the new Assure() definition is different from the existing/legacy Must() definition. New debugs() labels, new message syntax, new internal implementation.

I have several reasons:

 1) the current (mis)users of Must() were written assuming / relying on its existing behaviour and output.
 2) other followup PRs will be cleaning/removing bad uses of Must().
 3) linking the two definitions is polluting/degrading the new Assure() API implementation.
 3) linking the two definitions is polluting/degrading the old Must() log outputs.
 4) we do not need to change it

rousskov replied:
> > Alex: I see no valid reason for you to insist that Must() and Assure() implementations must duplicate each other.

> Amos: I am **not** insisting that. I am requesting that the existing Must() definition be left **untouched** by this PR.

... which will necessarily (poorly) duplicate Must() and Assure() implementations. On rare occasions, code duplication is the lesser of two evils, but I see no sufficient justification for implementing the same thing twice here:

* Must() implementation is check-condition-and-throw-exception code
* Assure() implementation is check-condition-and-throw-exception code

The only differences between the two correct implementations are the debug levels (0 vs. 3), and that the level-0 message needs a "ERROR: Squid BUG: " label. All the implementation code that is difficult to get right is actually identical.


> AFAICS, the new Assure() definition is different from the existing/legacy Must() definition. New debugs() labels, new message syntax, new internal implementation.

Yes, the message level and the message label/prefix naturally differ between Assure() and Must(), but that trivial difference can be (and is) expressed in one internal function parameter. We should not duplicate non-trivial internal code just because Assure() and Must() have different message levels (and because level-0 messages need a classification label).

> I have several reasons:

> 1. the current (mis)users of Must() were written assuming / relying on its existing behaviour and output.

This is not a valid reason because that old Must() behavior is preserved by this PR. 

Also, since branch commit 82817f6, the old level-0/1 output is also preserved. Here is a sample of the two failed Must(fail) checks that lead to level-0/1 messages, one in the official code and one in the PR branch code:


Before this PR:

```
2022/03/18 10:56:08.846| FATAL: check failed: false
    exception location: main.cc(1706) SquidMain
```

After PR commit 82817f6:

```
2022/03/23 17:37:15.791| FATAL: check failed: false
    exception location: main.cc(1661) SquidMain
```

Debugging (i.e. level-3) output has a different reporting location, of course, but is otherwise unchanged. Here is a sample:

Before this PR:

```
2022/03/18 10:56:08.846| 0,3| main.cc(1706) SquidMain: check failed: false
    exception location: main.cc(1706) SquidMain
```

After PR commit 82817f6:

```
2022/03/23 17:37:15.791| 0,3| Assure.cc(21) ReportAndThrow_: check failed: false
    exception location: main.cc(1661) SquidMain
```

Prior to branch commit 82817f6, the "check failed" prefix was not a part of the TextException text, but that accidental omission was my bug -- that text should be preserved, of course.


> 2. other followup PRs will be cleaning/removing bad uses of Must().

This is not a valid reason because this PR has no effect on cleaning/removing bad uses of Must(). When the very last Must() caller is gone, we will simplify Assure() implementation, but that change itself is simple, and I have already volunteered to do it.

> 3. linking the two definitions is polluting/degrading the new Assure() API implementation.

True, but having the level parameter and a simple extra condition (to label a level-0 message) is much better than poorly duplicating Must() and Assure() implementation, especially when that condition is not on a performance sensitive path and has a 5% binary size reduction side effect.

> 4. linking the two definitions is polluting/degrading the old Must() log outputs.

This is not a valid reason because this PR has no such effects on Must() log output.

> 5. we do not need to change it

This is not a valid reason because the need for that change has been established: Prevention of (poor) code duplication (and binary size reduction as a nice side effect of a better implementation).

In summary, we should not block a mishandled PR because it trades poor code duplication (and binary bloat) for an extra simple internal parameter/condition (outside performance sensitive paths).

rousskov reviewed (commented):
> I think we should go for a fully correct API definition for Assure() here without the hacks to allow compatibility with deprecated Must().

I agree. I believe there are no such API compatibility hacks in this PR.

I have applied the change you have insisted on. Please unblock this PR.

kinkie commented:
Can we recap the full spectrum of validation APIs and their intended main use cases?

I have:
- Assure()
- Must()
- assert()
- ??? for input validation

rousskov commented:
> Can we recap the full spectrum of validation APIs and their intended main use cases?

Sure:

* assert(): code logic check; legacy callers and callers that think the entire process (at least) should be killed if the condition is violated
* Assure(): code logic check; the caller thinks there is no reason to kill the entire process if the condition is violated; just kill the current transaction/component
* Must(): Deprecated. Currently used as either Assure() or input validation. Manual analysis required to distinguish between those two typical use cases. Some existing Must()s should even be converted to assert()s!
* Input validation should be done via context-specific code rather than some generic one-size-fits-all API

yadij commented on src/base/Assure.h:
```diff
@@ -0,0 +1,52 @@
+/// Calls ReportAndThrow() if needed. Reduces caller code duplication.
+/// Do not call directly; use Assure() instead.
+/// \param description c-string explaining the condition (i.e. what MUST happen)
+#define Assure_(debugLevel, condition, description, location) \
```
> but this logic does not work in reverse: Since ReportAndThrow_() is reused by non-~asserting~assuring code, that function should not and cannot be subject to `NDEBUG`.

Without Must() changes in this PR there is no such "non-~asserting~assuring code". The reverse logic works fine.

rousskov replied:
> Without Must() changes ...

Must() changes exist and are discussed in another change request. AFAICT, you tried to make a case that this change request is based on a _self-contained_ logic/principle/truth (rather than dependent on another change request resolution). That did not work. If you actually meant that this change request is simply dependent on a particular resolution of the other change request, then I agree, and (still) consider this change request addressed (because that particular resolution has been rejected).

Please unblock this PR.

yadij commented on src/base/Assure.cc:
```diff
@@ -0,0 +1,25 @@
+ReportAndThrow_(const int debugLevel, const char *description, const SourceLocation &location)
+{
+    const TextException ex(description, location);
+    const auto label = debugLevel <= DBG_IMPORTANT ?
+                       "ERROR: Squid BUG: assurance failed" : "check failed";
+    // TODO: Consider also printing the number of BUGs reported so far. It would
+    // require GC, but we could even print the number of same-location reports.
+    debugs(0, debugLevel, label << ": " << ex);
```
> > all Assure() failures are bugs
> 
> True, but most cache.log readers cannot tell whether the logged `ERROR` is the result of an Assure() failure.

False, all Assure()'s state "assurance failure" other log entries do not.

> And most readers do not (and should not) know what Assure() is or what special meaning we assign to "assurance".

... a problem introduced by this PR creating a novel alternative to assert. Which users are already familiar with.


> They do (or should) know what a _bug_ is though.

They know that "ERROR" means some sort of a bug.

> Since these code logic check failures are reported using non-standard "assurance" terminology, we should add a standard term like "bug" to explain what is going on.
> 

No. We should document what "assurance failure" means **somewhere**. So admin seeing it for the first time can look it up. That would be the wiki page documenting cache.log ERROR meanings.

That somewhere should not be polluting every related log line. Like I said, we already have "ERROR" to indicate type+severity of the problem.

For example, I bet you can tell me whether this is a bug or not
 ```error:14094438:SSL routines:ssl3_read_bytes:tlsv1 alert internal error (http_request_failed)```

yet nowhere does the word "BUG" appear in that log entry.

yadij replied:
> > They know that "ERROR" means some sort of a bug.
> 
> ERROR is usually not a bug. Very few ERROR messages are some sort of a bug.
> 

That itself would be a bug / badly written message.  From the FAQ definition:  ```a serious problem which has broken an individual client transaction and may have some effect on other clients indirectly. But has not completely aborted all traffic service```

Problems which are not internal errors for Squid, eg caused by external things should be WARNING (or just notices). FATAL and ERROR are for issues which have killed all or part of Squid service - aka "internal errors".

rousskov replied:
> > ERROR is usually not a bug. Very few ERROR messages are some sort of a bug.
 
> That itself would be a bug / badly written message. 

I see no reason why it would but that is because your definition of `ERROR` is different from mine.

> From the FAQ definition: `a serious problem which has broken an individual client transaction and may have some effect on other clients indirectly. But has not completely aborted all traffic service`

The quoted phrase does not contradict what I said AFAICT.

N.B. Please do not misinterpret this response as an implicit agreement that we should be using the above "FAQ definition". I am just saying that even if we do use it, my statement remains correct.


> Problems which are not internal errors for Squid (eg caused by external things) should be WARNING or just notices. FATAL and ERROR are for issues which have killed all or part of Squid service - aka "internal errors".

I (and the majority of Squid code) disagree with that opinion.

Errors are things that went wrong. Warnings are notices, advice, or predictions about something that may go wrong in the future. These are natural, widely-understood concepts. There is no need to redefine them.

We should be using ERROR and FATAL labels for non-fatal and fatal (to the current process) errors. We should be using WARNING for warnings.

kinkie commented:
It's been a month with no updates on this PR and I'd like to resurrect it
There are still some differences of opinion, but in the spirit of done is better than perfect, I'd recommend to just merge it and move on to make use of it

Any objection

rousskov commented:
> It's been a month with no updates on this PR 

To be more precise, it has been 23 days since https://github.com/squid-cache/squid/pull/864#discussion_r833771546 but, for a `review-1` PR that has been mishandled for a very long time before that, it is a long delay indeed.

> I'd like to resurrect it There are still some differences of opinion, but in the spirit of done is better than perfect, I'd recommend to just merge it and move on to make use of it
> 
> Can we?

I hope we can, but it is in Amos hands. I did my best to [detail](https://github.com/squid-cache/squid/pull/864#discussion_r833771546) how this PR "trades poor code duplication (and binary bloat) for an extra simple internal parameter/condition (outside performance sensitive paths)". IMO, we should gladly accept that trade rather than demand changes that would implement the opposite choice (i.e. poorly duplicate code and preserve binary bloat). FWIW, if I were a PR reviewer, I would have either accepted the opposite choice (because it _can_ be considered as valid due to Must-related excuses) or improved the PR to remove duplication and bloat (because those are obviously bad things).

There were many instances where the new code should but could not use Assure().

yadij commented:
Sure. Commit this and I will followup with the PR removing redundant #define wrappers now that Must() == Assure().

rousskov commented:
> Sure. Commit this 

@yadij, I cannot "commit this" because your are still blocking this PR.

> I will followup with the PR removing redundant #define wrappers now that Must() == Assure().

There must be some misunderstanding here because while both APIs share the same basic implementation, Must() behavior is different from Assure() behavior and that essential/meaningful difference cannot be removed. Said that, we can postpone that discussion to another PR.

yadij commented:
> FWIW, if I were a PR reviewer, I would have either accepted the opposite choice (because it _can_ be considered as valid due to Must-related excuses) or improved the PR to remove duplication and bloat (because those are obviously bad things).
> 

So ... whenever I am convinced you are wrong about something (everyone believes their own opinion/way is "valid"), I should just step in as reviewer and re-write your code to the "valid" implementation. Excellent point of view there.

....  I hope you can see why it is considered bad practice for a *reviewer* to do that.

yadij reviewed (commented):
Removing the blocking review. Though note that all outstanding issues remain ... Assure() is still automatically mis-used in *all* places where Must() was previously mis-used.

yadij commented:
> > I will followup with the PR removing redundant #define wrappers now that Must() == Assure().
> 
> There must be some misunderstanding here because, while both APIs share the same basic implementation, Must() behavior is different from Assure() behavior 

How do you think they are different when both are wrapper #define's pointing to the same function?

rousskov commented:
> Removing the blocking review.

Still blocked. FYI: You need to _dismiss_ the old negative review, not post a neutral comment.

> Though note that all outstanding issues remain ...

> Assure() is still automatically mis-used in all places where Must() was previously mis-used.

Assure() is not used at all! Future PRs will use Assure(). Hopefully, they will do it correctly, causing no misuse.

> NDEBUG will now disable input validation in all places relying on Must() exceptions to prevent input corruption

`NDEBUG` does not affect Must()s. This PR does not change to how Must()s are (ab)used for input validation.

>> Alex: while both APIs share the same basic implementation, Must() behavior is different from Assure() behavior

> How do you think they are different when both are wrapper #define's pointing to the same function?

Both point to the same macro, but give that macro different parameters. Here is the same information expressed as a diff:

```diff
- Assure_(0, (condition), ("assurance failed: " description), ...) // <--- Assure()
+ Assure_(3, (condition), ("check failed: " description), ...) // <--- Must()
```

rousskov commented:
> > FWIW, if I were a PR reviewer, I would have either accepted the opposite choice (because it _can_ be considered as valid due to Must-related excuses) or improved the PR to remove duplication and bloat (because those are obviously bad things).

Given your negative reaction to the above neutral comment, I am guessing I should have explicitly said "or improved (_with author permission_)", to avoid a misinterpretation of my comment as a blueprint for hostile PR takeovers!

> So ... whenever I am convinced you are wrong about something [...], I should just step in as reviewer and re-write your code to the "valid" implementation. Excellent point of view there.

Please note that my comment was not really about a _rewrite_, but the next logical step in PR evolution -- using the already written by the PR author function (with, hopefully, minor adjustments like adding a simple parameter) to replace a problematic implementation of an old API.

> .... I hope you can see why it is considered bad practice for a _reviewer_ to do that.

Hostile actions towards PR author branches are obviously bad practice, but there is nothing wrong with a reviewer offering to improve the PR further when, for example, the author is busy with something else or those improvements are easier to implement than describe with sufficient detail. Adding a few commits to author's PR branch can save a lot of time as long as everybody involved is careful, respectful, mindful or potential conflicts, etc. This is one of the reasons GitHub enables such collaboration among authors and qualified reviewers _by default_.
